pkg/collectors/log_follower.go

```diff
@@ -0,0 +1,247 @@
+	pollInterval       int
+	lpLock             sync.RWMutex
+	withTimeStamps     bool
+	running            bool
+	pruned             bool
+}
+
+const (
+	LogsCollectorName = "Logs"
+	LogsInfo          = "log-line"
+)
+
+func (logs *LogsCollector) GetPollInterval() int {
+	return logs.pollInterval
+}
+
+func (logs *LogsCollector) IsAnnouncer() bool {
+	return false
+}
+
+func (logs *LogsCollector) SetLastPool() {
+	if !logs.lpLock.TryLock() {
+		return
+	}
+	defer logs.lpLock.Unlock()
+	logs.lastPoll = time.Now()
+}
+
+// Start sets up the collector so it is ready to be polled
+func (logs *LogsCollector) Start() error {
+	go logs.writeToLogFile()
+	logs.running = true
+	return nil
+}
+
+func (logs *LogsCollector) consumeLine(line string, writer io.StringWriter) {
+	splitLine := strings.SplitN(line, " ", 2) //nolint:gomnd // moving this to a var would make the code less clear
+	if len(splitLine) < 2 {                   //nolint:gomnd // moving this to a var would make the code less clear
+		return
+	}
+	timestampPart := splitLine[0]
+	lineContent := splitLine[1]
+	timestamp, err := time.Parse(time.RFC3339, timestampPart)
+	if err != nil {
+		// This is not a value line something went wrong
+		return
+	}
+
+	if seen, ok := logs.seenLines[timestamp]; ok {
+		for _, seenContent := range seen {
+			if seenContent == lineContent {
+				// Already have this line skip it
+				return
+			}
+		}
+	}
+	logs.seenLines[timestamp] = append(logs.seenLines[timestamp], lineContent)
+	logs.pruned = false
+	if logs.withTimeStamps {
+		_, err = writer.WriteString(line + "\n")
+	} else {
+		_, err = writer.WriteString(lineContent + "\n")
+	}
+	if err != nil {
+		log.Error(fmt.Errorf("failed to write log output to file"))
+	}
+}
+
+func (logs *LogsCollector) writeToLogFile() {
+	logs.wg.Add(1)
+	defer logs.wg.Done()
+
+	fileHandle, err := os.OpenFile(logs.logsOutputFileName, os.O_CREATE|os.O_WRONLY, logFilePermissions)
+	utils.IfErrorExitOrPanic(err)
+	defer fileHandle.Close()
+
+	for {
+		select {
+		case <-logs.quit:
+			// Consume the rest of the lines so we don't miss lines
+			for len(logs.lines) > 0 {
+				line := <-logs.lines
+				logs.consumeLine(line, fileHandle)
+			}
+			return
+		case line := <-logs.lines:
+			logs.consumeLine(line, fileHandle)
+		default:
+			if !logs.pruned {
+				log.Info("pruning seen lines")
+				for key := range logs.seenLines {
+					if time.Since(key) > keepDuration {
+						delete(logs.seenLines, key)
+					}
+				}
+				logs.pruned = true
+			} else {
+				time.Sleep(time.Microsecond)
+			}
+		}
+	}
+}
+
+func (logs *LogsCollector) processLines(line string) string {
+	if strings.ContainsRune(line, lineDelim) {
+		lines := strings.Split(line, "\n")
+		for n := 0; n < len(lines)-2; n++ {
+			logs.lines <- lines[n]
+		}
+		line = lines[len(lines)-1]
+	}
+	return line
+}
+
+func (logs *LogsCollector) poll() error {
+	logs.lpLock.RLock()
+	defer logs.lpLock.RUnlock()
+	sinceSeconds := int64(time.Since(logs.lastPoll).Seconds())
```

Comment:
move the `sinceSeconds` calculation to as late as possible before the `SetLastPoll` method to avoid delays from network issues when resolving the Pod Name.